### PR TITLE
static assert that bundles do not have duplicate types

### LINKED ISSUE #2387: Enforce type-uniqueness of Bundles
## What problem does this solve or what need does it fill?

Currently, bundles can have fields of duplicate types. For example:

```rs
use bevy::prelude::*;

struct Foo(u32);

#[derive(Bundle)
struct Bar{
  foo1: Foo,
  foo2: Foo,
}
```
compiles with no warning.

However, this behaviour is deeply wrong, and leads to silent, subtle bugs in end user code.
Each field in a bundle represents one component that should be inserted into the final entity, and entities can only have one unique component of each (Rust) type.

Any duplicate components will have their data silently erased by the last field of the same type; wasting work and causing subtle and serious logic bugs. In the demonstrated example, setting the value of `foo1` has no effect: it will always be overwritten by `foo2`.

This problem is particularly rough when working with type aliases (where the problem is not immediately obvious) and nested bundles (where there can be a large number of elided types, and new errors can occur without any change to user code on a new release if more components are added).

## What solution would you like?

The `#[derive(Bundle)` macro should check for duplicated types, and error (ideally at compile time) if they are found with an error message that displays both the field and type names.

Be sure to test that this check works with:
1. type aliases
2. nested bundles

## What alternative(s) have you considered?

1. This could be enforced at component-insertion time, using #2241. I would prefer to leave that as an additional layer of opt-in safety, and instead provide compile time checks for obvious mistakes.
2. We may want a way to opt-out of this; perhaps with an #[allow_duplicate] attribute on specific fields.
3. An error, especially at compile time, is better than a warning, as it will either be immediately fixed or permanently ignored.
4. It is by far more important that this is enforced for the Bundle derive macro, but it might be nice to enforce it on manual impls too.

### PR DESCRIPTION
**Pull request by NathanSWard**
# Objective

- Currently it is possible for a user to define a bundle that contains duplicate types. However, this is incorrect and can lead to various bugs.
- Partially addresses #2387

## Solution

- Implement a static check for Bundles that when using `#[derive(Bundle)]` will fail to compile if any of the inner types are duplicates.

### CODE REVIEW

**Comment by NathanSWard** [comment]
As a quite note, this is sort of a first pass solution. 
I don't believe this handles nested bundles yet, however, I just wanted to see if this is the general direction we want to go with this issue.

File: crates/bevy_ecs/macros/src/lib.rs [hunk 1]
```diff
@@ -165,6 +178,19 @@ pub fn derive_bundle(input: TokenStream) -> TokenStream {
                 #(#field_get_components)*
             }
         }
+
+        #[cfg(test)]
```
Comment (MinerSebas) [inline crates/bevy_ecs/macros/src/lib.rs:182]: Why is the `#[cfg(test)]` (neccessary) here?
Shouldn't that check always happen, regardless of whether you run tests?

Comment (MinerSebas) [reply crates/bevy_ecs/macros/src/lib.rs:182]: Looking at the CI Failure, the added doc-test didn't fail to compile, because of the `#[cfg(test)]`.

Comment (NathanSWard) [reply crates/bevy_ecs/macros/src/lib.rs:182]: Ideally, I want to prevent any code-gen when you're actually building a binary.

Comment (NathanSWard) [reply crates/bevy_ecs/macros/src/lib.rs:182]: Though it may not be necessary... 🤔

Comment (NathanSWard) [reply crates/bevy_ecs/macros/src/lib.rs]: Nvm, you're correct, we should remove this otherwise you'll get inconsistent behavior haha.

**Comment by bjorn3** [comment]
Another option would be to have `trait AssertDistinctComponentsMyBundle {}` and then an impl of this trait for each field type. That will result in a conflicting trait impl if the types are equal.

**Comment by NathanSWard** [comment]
> Another option would be to have trait AssertDistinctComponentsMyBundle {} and then an impl of this trait for each field type. That will result in a conflicting trait impl if the types are equal.

This is how the current implementation is conducted... 🙃 (I'm pretty sure?)

**Comment by bjorn3** [comment]
Oops, I thought it used a trait method and then error on ambiguity or something like that.

**Comment by NathanSWard** [comment]
This should now also work with nested bundles!! :)

**Comment by NathanSWard** [comment]
Currently this implementation only support a single layer of checking.
e.g.
```rust
#[derive(Bundle)]
struct A {
    a: usize,
    b: usize,
}
```
will fail to compile because `a` and `b` are in the same structure.

However,
```rust
#[derive(Bundle)]
struct A {
    a: usize,
}

#[derive(Bundle)]
struct A {
    #[bundle]
    a: A,
    b: usize,
}
```
will still compile since you cannot implement traits for foreign types...

I still think this addition is a nice QOL improvement, however we should keep an issue open to track nested bundle static checking.

**Comment by NathanSWard** [comment]
@TheRawMeatball I'm curious what your thoughts are with this issue.
From what we've seen, it's not possible to assert at compile time that bundles are correct (due to nested bundles).
However is this current PR even worth pursuing? 
As in should we support static checking for a single layer or not support it until we can find a solution for nested bundles.

**Comment by alice-i-cecile** [comment]
IMO failing if we *know* something is incorrect is worth doing, even if we can't catch every case.

**Comment by mockersf** [comment]
Going on a tangent about nested bundle: is the possibility to do it really worth it? I'm not completely convinced. The issues raised in #1588  and #2331 when trying to use it are valid, and if we don't even use it in standard Bevy I would argue this should be removed.

**Comment by TheRawMeatball** [comment]
I'd strongly prefer nested bundles over fully static assertions personally. They're very useful for spawning more complex entities.

**Comment by mockersf** [comment]
```rust
#[derive(Bundle, Debug)]
pub struct Position {
    pub x: u8,
    pub y: u8,
}
```

gives:
```
error[E0119]: conflicting implementations of trait `static_assert_Position_does_not_have_duplicate_types::AssertDistinctComponentsForPosition` for type `u8`
  --> src/main.rs:46:10
   |
46 | #[derive(Bundle, Debug)]
   |          ^^^^^^
   |          |
   |          first implementation here
   |          conflicting implementation for `u8`
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

**Comment by TheRawMeatball** [comment]
Looks pretty good to me! Hopefully in the far future when more rust features are stable we can make it work for nested stuff too, but for now this seems like a useful compromise.

**Review by mockersf — APPROVED** [review]

**Comment by NathanSWard** [comment]
I just realized, this impl may not work with generic bundles....

**Comment by NathanSWard** [comment]
There's a few problems with the PR, including generic bundles, and not being able to recurse and check nested bundles.
I'm going to close this for now, and if I can find a more viable solution I'll open an new PR accordingly.

**Pull request closed without merging**